Refactor for multithread

td-extreme opened the pull request.

td-extreme committed:
Refactor to prepare for transition to multithread

client socket is now open and closed inside the loop in the main class.

Removed the HttpReaderWriter class.

HttpRequestBuilder class now uses the HttpRequestReader and
HttpRequestParser class instead of the HttpReaderWriter giving the
HttpRequestBuilder a String and asking it to parse and build a request.

td-extreme committed:
corrected HttpServerRunner to use interface

td-extreme committed:
add new UML class diagram

dadarek commented on src/main/java/HttpServerRunner.java:
```diff
@@ -3,29 +3,33 @@
+    catch (InvalidHttpRequestException e) {
+      handler = handlerRouter.selectHandlerBadRequest();
+    }
+    catch (IOException e) {
```
This feels like a low-level exception bubbled up. Also feels like it _should_ be handled.

td-extreme replied:
Do you mean it should be handled lower (ie in the `SocketIO` or `HttpRequestBuilder`), or that there should be logic in the catch block to close the socket and finish any further execution of code in the thread?

dadarek replied:
Where it should be handled and how is up to you - but my point is that the reason it's being thrown here is because `httpRequestBuilder` or `handlerRouter` depend on something that throw it–they don't actually throw it themselves.

By letting it bubble up you're sort of exposing that and forcing `HttpServerRunner` to also depend on the "details" of those two classes it uses.

dadarek commented on src/main/java/HttpServerRunner.java:
```diff
@@ -3,29 +3,33 @@
+    try {
+      httpResponseWriter.sendHttpResponse(client, response);
+    }
+    catch (IOException e) { }
```
Ditto. We should probably at least make sure to close the client if we can't write back to it.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -1,41 +1,27 @@
 import java.io.IOException;
 
 public class HttpRequestBuilder implements IRequestBuilder {
+  IRequestReader reader;
```
Are these private by default? Or package-protected?

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -1,41 +1,27 @@
   }
 
-  public HttpRequest createRequest(String rawRequest) throws InvalidHttpRequestException {
+  public HttpRequest createRequest(IMessageIO client) throws InvalidHttpRequestException, IOException {
```
This is really more like `getNextRequest`, right? It doesn't just build it out of nothing.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -1,41 +1,27 @@
     HttpRequest request;
-    String requestLine = parseRequestLine(rawRequest);
-    HashMap<String, String> headers = parseHeaders(rawRequest);
+    String rawRequest = reader.getMessage(client);
```
What does `getMessage` actually mean? It's not obvious what this method does just by the name.

td-extreme replied:
I'll come up with a better name for this.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -1,41 +1,27 @@
-    HashMap<String, String> headers = parseHeaders(rawRequest);
+    String rawRequest = reader.getMessage(client);
+    String requestLine = parser.parseRequestLine(rawRequest);
+    HashMap<String, String> headers = parser.parseHeaders(rawRequest);
```
What is the benefit of having to call `parseRequestLine` followed by `parseHeaders` on a `RequestParser`, and then putting those together yourself to create an HttpRequest`?

td-extreme replied:
Would it be okay for the `HttpRequestBuilder` to Contain the parsing logic and also be injected with a `SocketIO` that is asks to get a String and Bytes(if needed, because there is a body)??? 
Not sure if those are two different responsibilities. I don't think they are..

dadarek replied:
Not sure about the `SocketIO` injection - that _may_ be overkill? But I can certainly see the parser just exposing one fn that takes the `rawRequest` and returns a built-up `HttpRequest` (possibly without the body?).

dadarek commented on src/main/java/request/HttpRequestParser.java:
```diff
@@ -0,0 +1,31 @@
+
+  public String parseRequestLine(String rawRequest) throws InvalidHttpRequestException {
+    String requestLine = rawRequest.split("\r?\n")[0];
+    if (!requestVerifier.isValid(requestLine)) {
```
Hmmm... I wonder if this `requestVerifier` should still be injected. Maybe it could be pulled into this class? Or at least, maybe it can be just used directly by this class, no need to inject it?

dadarek commented on src/main/java/request/HttpRequestParser.java:
```diff
@@ -0,0 +1,31 @@
+    return requestLine;
+  }
+
+  public HashMap<String, String> parseHeaders(String rawRequest) throws InvalidHttpRequestException  {
```
Where does this method actually throw the `InvalidHttpRequestException`?

dadarek commented on src/main/java/request/HttpRequestReader.java:
```diff
@@ -0,0 +1,13 @@
+package com.td.HttpServer;
+import java.io.IOException;
+
+public class HttpRequestReader implements IRequestReader {
```
Feels like this class does nothing more than delegate to IMessageClient. What's the purpose for it?

td-extreme replied:
I suppose the `HttpRequestBuilder` could just use the `SocketIO` class directly.

dadarek replied:
Yeah, I don't see any value to this class as it is. It just delegates directly to a class/interface that your code still owns.

dadarek commented on src/main/java/response/HttpResponseWriter.java:
```diff
@@ -0,0 +1,20 @@
+package com.td.HttpServer;
+import java.io.IOException;
+
+public class HttpResponseWriter {
```
Same thing here. This class seems like a very simple delegator. Can you justify it's existence?

td-extreme replied:
As is I can't justify it. It would probably be okay to get ride of the `HttpResponseFormatter` class and put that logic into this class, though. So that this class would format the response and then send it to the `SocketIO`

dadarek replied:
Yeah, if it's simple delegation it's hard to justify _unless_ you're delegating to things you don't own (i.e. file system calls, external service calls, other-library calls).

dadarek commented on src/test/java/HttpRequestBuilderTest.java:
```diff
@@ -0,0 +1,40 @@
+    try {
+    builder.createRequest(new DummyIO());
+    }
+    catch (InvalidHttpRequestException e) {}
```
These seem a little clunky. Is there a way to avoid this or clean it up?

td-extreme replied:
This exception shouldn't get thrown during the test and if it does it clearly means the test failed. With that in mind would it be okay to just have the test method throw the exception, thus removing the try catch block??

dadarek replied:
Probably. If a test method throws an exception that will fail the test. That's probably a fine way to fail a tests :) Not sure if everyone shares that opinion.

dadarek commented on src/test/java/HttpRequestParserTest.java:
```diff
@@ -0,0 +1,38 @@
+    try {
+      map = parser.parseHeaders(rawRequest);
+    }
+    catch (InvalidHttpRequestException e) {}
```
Is `InvalidHttpRequestException` a "feature" or functionality or specification of HttpRequestParser?

dadarek commented:
class diagram seems to be a little out of date.

Specifically, the `IFiloIO` interface has more methods than actually appear on it (BTW, will there be a PR to clean up the interfaces? I forgot what we decided on Friday.)

dadarek commented:
Also, re: class diagram - the handlers' dependencies don't look correct.

Take a closer look at it to make sure it's accurate.

dadarek commented:
One last thing. Consider a few alternatives for what we could be doing with this proliferous `HttpRequest` class that's hijacking the codebase. Maybe a good blog. May also be a good topic for soliciting feedback from craftsman around the office.

Try hitting up Paul Pagel if he's in the office this week. Maybe Colin Jones (which I know he's probably in Libertyville) but could probably slack him and just send him the UML diagram once you get it up to date.

Then let's talk about what you came up with.

td-extreme committed:
Merge HttpResposneFormatter & HttpResposneWriter

td-extreme committed:
added new class diagram

td-extreme committed:
Reorganize classes and method names

Removed
    HttpResponseReader
Renamed
    IMessageIO interface to IClientSocketIO
    SocketIO to ClientSocketIO
    MockSocketIO to MockClientSocketIO

td-extreme committed:
HttpVerififer no longer injected HttpRequestParser.

td-extreme committed:
Added BadConnectionException

td-extreme committed:
updated class diagram

td-extreme committed:
Added printStackTrace()

dadarek commented on src/main/java/HttpServer.java:
```diff
@@ -7,16 +7,21 @@
 
   public static void main(String[] args) throws IOException {
```
Does main still actually throw an IOException?

td-extreme replied:
Not anymore. Totally missed that.

td-extreme replied:
Actually, it still can. Because of the ServerSocket being open. Not sure if that should be caught. I mean if the program can't open the ServerSocket then it really can't do anything else.

dadarek replied:
Can you recommend something?

td-extreme replied:
Well, I would think if the Port can't be opened you would want to output to the user that the port can not be open. Right now it will output the stack trace with that error, maybe instead the exception could be caught so that I could make the error message look nice instead of a stack trace.

dadarek replied:
That would probably make for a better user experience.

I was also thinking about what we talked about today: IOException comes from the lower-level components, yet it's being bubbled up through their interfaces.

dadarek commented on src/test/java/MockClientSocketIO.java:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  public void sendString(String rawString) {}
+  public void openClientConnection() {}
```
This is a sign that your interfaces are too generic and _probably_ owned by multiple classes.

dadarek commented on src/main/java/response/HttpResponseWriter.java:
```diff
@@ -1,18 +1,28 @@
+    }
+    catch (IOException e) {
+      e.printStackTrace();
+      throw new BadConnectionException();
```
One last thing about exceptions - it's a good idea to wrap exceptions so that, in case you need do need it (because you're debugging or reading error logs, _not_ because the layer above you wants to get it's hands on it programmatically).

`new BadConnectionException(e)`

That _should_ work out of the box because deriving from `Exception` will get you those constructors free of charge.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -1,41 +1,28 @@
       }
+      return request;
+    }
+    catch (IOException e) {
```
Remember the thing about scoping your catches to cover only relevant blocks?

dadarek replied:
That'll also help clean this method up a little bit. Few things going on here ...

dadarek commented on src/main/java/HttpServerRunner.java:
```diff
@@ -3,29 +3,40 @@
 
-  public void runServer() throws IOException {
-    HttpRequest request;
+  public void run(ClientSocketIO client) {
```
Method getting a little chunky now ...

dadarek commented:
+1

td-extreme committed:
try block only around exception throwing code.

td-extreme committed:
Wrapped throwing of BadConnectionException

td-extreme committed:
Broke up IClientSocketIO interface

There are now three seperate interfaces
  IClientSocketOpenClose
  IClientSocketInput
  IClientScoketOutput

Also broke up the MockClientSocketIO class into
  MockClientSocketInput
  MockClientSocketOutput

td-extreme committed:
Removed no longer used Mock

td-extreme committed:
Broke run method up

td-extreme committed:
Fixed visibility modifiers.

td-extreme committed:
Add exception handling to main.

mikeknep commented:
👍
